plugins/widget/plugin.js

```diff
@@ -3310,70 +3327,268 @@
+			widget.initEditable( editableName, typeof editableDef == 'string' ? { selector: editableDef } : editableDef );
+		}
+	}
+
+	function setupMask( widget ) {
+		if ( widget.mask === true ) {
+			setupFullMask( widget );
+		} else if ( widget.mask ) {
+			// Buffer to limit number of separate calls to 'setupPartialMask', e.g. during writing.
+			var maskBuffer = new CKEDITOR.tools.buffers.throttle( 250, setupPartialMask, widget ),
+				changeListener,
+				blurListener;
+
+			widget.on( 'focus', function() {
+				changeListener = widget.editor.on( 'change', maskBuffer.input );
+				blurListener = widget.on( 'blur', function() {
+					changeListener.removeListener();
+					blurListener.removeListener();
+				} );
+			} );
+
+			// Focusing editable doesn't trigger focus on widget, so listen to those events separately.
+			for ( var editable in widget.editables ) {
+				widget.editables[ editable ].on( 'focus', function() {
+					widget.editor.on( 'change', maskBuffer.input );
+					// If widget was focused before focusing editable, the 'blur' event has to be removed.
+					// Otherwise on Chrome it will trigger after the focus event and cancel listening to
+					// changes (on FF it works inversely).
+					if ( blurListener ) {
+						blurListener.removeListener();
+					}
+				} );
+				widget.editables[ editable ].on( 'blur', function() {
+					widget.editor.removeListener( 'change', maskBuffer.input );
+				} );
 			}
+			maskBuffer.input();
 
-			var label = this.getLabel ? this.getLabel() : getLabelDefault.call( this );
+			// FF renders image-like widget very late (much after 'instanceReady') so mask creation
+			// has to be postponed. The same is true for switching between WYSIWYG and source mode,
+			// just different event and smaller delay are needed.
+			if ( CKEDITOR.env.gecko ) {
+				widget.editor.on( 'instanceReady', function() {
+					setTimeout( function() {
+						maskBuffer.input();
+					}, 600 );
+				} );
+				widget.editor.on( 'mode', function() {
+					setTimeout( function() {
+						maskBuffer.input();
+					}, 100 );
+				} );
+			}
+		}
+	}
 
-			widget.wrapper.setAttribute( 'role', 'region' );
-			widget.wrapper.setAttribute( 'aria-label', label );
-		}, null, null, 9999 );
+	function setupFullMask( widget ) {
+		// Reuse mask if already exists (https://dev.ckeditor.com/ticket/11281).
+		var img = widget.wrapper.findOne( '.cke_widget_mask' );
+
+		if ( !img ) {
+			img = new CKEDITOR.dom.element( 'img', widget.editor.document );
+			img.setAttributes( {
+				src: CKEDITOR.tools.transparentImageData,
+				'class': 'cke_reset cke_widget_mask'
+			} );
+			widget.wrapper.append( img );
+		}
+
+		widget.mask = img;
+	}
+
+	function setupPartialMask() {
+		if ( !this.wrapper ) {
+			return;
+		}
+
+		// Original value of 'widget.mask' is substituted with actual mask element, so
+		// 'widget.maskPart' property was added to be able to adjust partial mask e.g. after resizing.
+		this.maskPart = this.maskPart || this.mask;
```

Comment:
I was thinking about it too and wanted to drop this for both flows, but I noticed that we have a test checking if mask is DOM element - but this may be both a business use case or just a way to determine if mask was created. What was first - 🥚 or 🐔? I have no clue, so decided to let it be as it is. However I'm open-minded here.